src/django/api/tests.py

```diff
@@ -259,3 +265,98 @@ def test_generated_token_is_not_deleted_on_logout(self):
         self.assertFalse(user.is_authenticated)
         token = Token.objects.filter(user=self.user)
         self.assertEqual(token.count(), 1)
+
+
+class GeocodingUtilsTest(TestCase):
+    def setUp(self):
+        settings.GOOGLE_GEOCODING_API_KEY = "world"
+
+    def test_geocoding_api_url_is_created_correctly(self):
+        self.assertEqual(
+            create_geocoding_api_url("hello", "US"),
+            (
+                "https://maps.googleapis.com/maps/api/geocode/json"
+                "?components=country:US&address=hello&key=world"
+            )
+        )
+
+    def test_geocoded_address_data_is_formatted_correctly(self):
+        formatted_data = format_geocoded_address_data(
+            parsed_city_hall_data['full_response'])
+        self.assertEqual(formatted_data, parsed_city_hall_data)
+
+
+class GeocodingTest(TestCase):
+    def test_correct_address_is_geocoded_properly(self):
+        geocoded_data = geocode_address('City Hall, Philly', 'US')
+        self.assertDictEqual(geocoded_data, parsed_city_hall_data)
+
+    def test_ungeocodable_address_returns_value_error(self):
+        with self.assertRaises(ValueError) as cm:
+            geocode_address('hello world', '$#')
+
+        self.assertEqual(cm.exception.args, ('No results were found',))
+
+
+class FacilityListItemGeocodingTest(TestCase):
+    def test_invalid_argument_raises_error(self):
+        with self.assertRaises(ValueError) as cm:
+            geocode_facility_list_item("hello")
+
+        self.assertEqual(
+            cm.exception.args,
+            ('Argument must be a FacilityListItem',)
+        )
+
+    def test_unparsed_item_raises_error(self):
+        facility_list = FacilityList(header='address,country,name')
+        item = FacilityListItem(
+            raw_data='1400 JFK Blvd, Philly,us,Shirts!',
+            facility_list=facility_list
+        )
+
+        with self.assertRaises(ValueError) as cm:
+            geocode_facility_list_item(item)
```

Comment:
Today I learned (or possibly relearned) that the binding of a `with` block stays in scope after the block completes. Good to know.

Reply:
I saw this in the Python tests documentation, which seemed like a nice way to test the type and result of an Exception!